File: leetcode/12/c++/solution.cpp
/*
    12. Integer to Roman
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    std::string intToRoman(int num) {
        uint num_i = 0;
        std::string roman_num = "";
        std::unordered_map<uint, std::string> roman_map = {
            {1, "I"},   {4, "IV"},   {5, "V"},   {9, "IX"},  {10, "X"},
            {40, "XL"}, {50, "L"},   {90, "XC"}, {100, "C"}, {400, "CD"},
            {500, "D"}, {900, "CM"}, {1000, "M"}};
        roman_num.reserve(10);
        for (uint i = 1; num > 0; i *= 10) {
            num_i = (num % 10);
            num /= 10;
            if (num_i == 1 || num_i == 4 || num_i == 5 || num_i == 9) {
                roman_num.insert(0, roman_map[num_i * i]);
            } else if (num_i > 0) {
                std::string temp_1 = roman_map[1 * i];
                for (size_t k = 0; k < num_i % 5; ++k) {
                    roman_num.insert(0, temp_1);
                }
                if (num_i > 5) {
                    roman_num.insert(0, roman_map[5 * i]);
                }
            }
        }
        return roman_num;
    }
};

File: leetcode/12/rust/solution.rs
/*
    12. Integer to Roman
    Author: Miguel Angel Bermeo Ayerbe
*/
impl Solution {
    pub fn int_to_roman(num: i32) -> String {
        let roman_map: Vec<(i32, &'static str)> = vec![
            (1000, "M"),
            (900, "CM"),
            (500, "D"),
            (400, "CD"),
            (100, "C"),
            (90, "XC"),
            (50, "L"),
            (40, "XL"),
            (10, "X"),
            (9, "IX"),
            (5, "V"),
            (4, "IV"),
            (1, "I"),
        ];
        let mut roman: String = String::new();
        let mut local_num = num;
        roman.reserve(10);
        for (roman_num, roman_str) in roman_map.iter() {
            while local_num >= *roman_num {
                roman.push_str(*roman_str);
                local_num -= *roman_num;
            }
        }
        roman
    }
}
